p2p/peer_behaviour.go

```diff
@@ -0,0 +1,140 @@
+package p2p
+
+import (
+	"errors"
+	"sync"
+)
+
+// ErrorPeerBehaviour are types of observable erroneous peer behaviours.
+type ErrorPeerBehaviour int
+
+const (
+	ErrorPeerBehaviourUnknown = iota
+	ErrorPeerBehaviourBadMessage
+	ErrorPeerBehaviourMessageOutofOrder
+)
+
+// GoodPeerBehaviour are types of observable positive peer behaviours.
+type GoodPeerBehaviour int
+
+const (
+	GoodPeerBehaviourVote = iota + 100
+	GoodPeerBehaviourBlockPart
+)
+
+// PeerBehaviour provides an interface for reactors to signal the behaviour
+// of peers synchronously to other components.
+type PeerBehaviour interface {
+	Behaved(peerID ID, reason GoodPeerBehaviour) error
+	Errored(peerID ID, reason ErrorPeerBehaviour) error
+}
+// SwitchPeerBehaviour relays peer behaviour signals to a Switch, which is
+// needed until Switch satisfies the PeerBehaviour interface.
+type SwitchPeerBehaviour struct {
+	sw *Switch
+}
+
+// Errored reports the ErrorPeerBehaviour of peer identified by ID to the Switch.
+func (spb *SwitchPeerBehaviour) Errored(peerID ID, reason ErrorPeerBehaviour) error {
+	peer := spb.sw.Peers().Get(peerID)
+	if peer == nil {
+		return errors.New("Peer not found")
+	}
+
+	spb.sw.StopPeerForError(peer, reason)
+	return nil
+}
+
+// Reports the GoodPeerBehaviour of peer identified by peerID to the Switch.
+func (spb *SwitchPeerBehaviour) Behaved(peerID ID, reason GoodPeerBehaviour) error {
+	peer := spb.sw.Peers().Get(peerID)
+	if peer == nil {
+		return errors.New("Peer not found")
+	}
+
+	spb.sw.MarkPeerAsGood(peer)
+	return nil
+}
+
+// Return a new switchPeerBehaviour instance which wraps the Switch.
+func NewSwitchPeerBehaviour(sw *Switch) *SwitchPeerBehaviour {
+	return &SwitchPeerBehaviour{
+		sw: sw,
+	}
+}
+
+// storedPeerBehaviour serves a mock concrete implementation of the
+// PeerBehaviour interface used in reactor tests to ensure reactors
+// produce the correct signals in manufactured scenarios.
+type StoredPeerBehaviour struct {
+	eb  map[ID][]ErrorPeerBehaviour
+	gb  map[ID][]GoodPeerBehaviour
+	mtx sync.RWMutex
```

Comment:
Our convention is to put mtxs immediately above the block of fields they protect